### PR TITLE
#183 object compositions

### LINKED ISSUE #183: context.py:117-121: Move context realization to pure to...
The puzzle `550-c08f412e` from #550 has to be resolved:

https://github.com/fidals/refarm-site/blob/471fe2d1d8b6867da6567126cd1cc6f242764887/catalog/context.py#L117-L121

The puzzle was created by duker33 on 17-Sep-18. 

Estimate: 120 minutes,     

If you have any technical questions, don't ask me, submit new tickets instead. The task will be "done" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### PR DESCRIPTION
**Pull request by artemiy312**
Closes #183

### CODE REVIEW

File: catalog/context/tags.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        return {
+            **super().context(),
+            'group_tags_pairs': self.qs().get_group_tags_pairs(),
+        }
+
+
+class Checked404Tags(Tags):
+
+    def qs(self):
+        tags = self.qs()
```
Comment (duker33) [inline catalog/context/tags.py]: `self.super.qs()` or smth. Now it's infinity recursion

File: catalog/models.py [hunk 2]
```diff
@@ -95,6 +95,7 @@ class ProductQuerySet(models.QuerySet):
     def get_offset(self, start, step):
         return self[start:start + step]
 
+    # @todo #183:60m Try to remove `ordering` arg from ProductQuerySet.get_by_category
```
Comment (duker33) [inline catalog/models.py]: it's about 30 min i think. We should move it from model to context.
And use ready context arch for it

File: catalog/context/products.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+from django.db.models import QuerySet
+
+from catalog.context.context import ModelContext
+from catalog.models import AbstractCategory
+
+
+class Products(ModelContext):
+
+    def __init__(self, qs: QuerySet):
```
[the remark this answers is not in the document]
Comment (artemiy312) [reply catalog/context/products.py]: The interface does not require it. A context implementation may have any constructor

[the remark this answers is not in the document]
Comment (duker33) [reply catalog/context/products.py]: @artemiy312 , but we should be able to chain any context to any.
They may have no common methods, but theirs dicts should be merged.
There is initial idea: merge dicts with simple chaining

File: catalog/context/context.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+
+class Contexts(Context):
+
+    def __init__(self, *contexts: typing.List[Context]):
+        self.contexts = contexts
+
+    def context(self):
+        return collections.ChainMap(
+            *[ctx.context() for ctx in self.contexts]
```
Comment (duker33) [inline catalog/context/context.py]:
```suggestion
            iter(ctx.context() for ctx in self.contexts)
```

File: catalog/context/products.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (duker33) [reply catalog/context/products.py]: @artemiy312 , y, it's much more good decision, then receiving contexts

File: catalog/context/context.py (hunk 4 above)
Comment (artemiy312) [reply catalog/context/context.py]: I don't see the point. How `iter` can help here? `ChainMap` takes variadic dict number as arguments, so i have to unpack contexts list

**Pull request merged**